Stripe/STPRedirectContext.m

```diff
@@ -48,16 +68,64 @@ - (void)dealloc {
     [self unsubscribeFromNotificationsAndDismissPresentedViewControllers];
 }
 
-- (void)startRedirectFlowFromViewController:(UIViewController *)presentingViewController {
+- (void)performAppRedirectIfPossibleWithCompletion:(STPNativeRedirectCompletionBlock)onCompletion {
     FAUXPAS_IGNORED_IN_METHOD(APIAvailability)
-    if ([SFSafariViewController class] != nil) {
-        [self startSafariViewControllerRedirectFlowFromViewController:presentingViewController];
+
+    if (self.state == STPRedirectContextStateNotStarted) {
+        NSURL *nativeUrl = [self nativeRedirectURLForSource:self.source];
+        if (!nativeUrl) {
+            onCompletion(NO);
+            return;
+        }
+
+        // Optimistically start listening in case we get app switched away.
+        // If the app switch fails we'll undo this later
+        _state = STPRedirectContextStateInProgress;
+        [self subscribeToUrlAndForegroundNotifications];
+
+        UIApplication *application = [UIApplication sharedApplication];
+        if ([application respondsToSelector:@selector(openURL:options:completionHandler:)]) {
+
+            WEAK(self);
+            [application openURL:nativeUrl options:@{} completionHandler:^(BOOL success) {
+                if (!success) {
+                    STRONG(self);
+                    self->_state = STPRedirectContextStateNotStarted;
```

Comment:
ooc, why `self->_state=` instead of just `_state=`?

Also, can't we just use the getters/setters in this method instead of ivars? I usually only use ivars in `init`, or if the getter/setter has side effects. Maybe we should add something to the style guide...

Reply:
You have to do `self->state` or you get a compiler warning about implicit retain of self.

Yeah there are some style guide stuff here I guess. I didn't actually make setters for these in the .m file, I generally prefer doing it this way, but we can change (And have a style guide conversation).

Reply:
oh right, forgot about that compiler warning. 👍